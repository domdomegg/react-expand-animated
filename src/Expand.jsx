import React, { Component } from 'react';
import PropTypes from 'prop-types';

const STATUS = {
  CLOSE: 'CLOSE',
  CLOSING: 'CLOSING',
  CLOSED: 'CLOSED',
  OPEN: 'OPEN',
  OPENING: 'OPENING',
  OPENED: 'OPENED',
};

class Expand extends Component {
  state = {
    status: this.props.open ? STATUS.OPEN : STATUS.CLOSE,
  };

  componentWillReceiveProps({ open }) {
    if (open !== this.props.open) {
      this.toggle(open);
    }
  }

  componentWillUnmount() {
    this.clearDelay();
  }

  getClientHeight = () => {
    if (this.refWrapper) {
      return this.refWrapper.scrollHeight;
    }

    return 'auto';
  };

  getExpandStyle = () => {
    const { status } = this.state;

    switch (status) {
      case STATUS.OPENING:
      case STATUS.CLOSE:
      case STATUS.CLOSED:
        return { height: 0, opacity: 0, overflow: 'hidden' };
      case STATUS.OPENED:
      case STATUS.CLOSING:
        return { height: this.getClientHeight(), opacity: 1, overflow: 'hidden' };
      default:
        return { height: 'auto', opacity: 1, overflow: 'unset' };
    }
  };

  getStyle() {
    const { duration, easing } = this.props;

    const transition = `height ${duration}ms ${easing}, opacity ${duration}ms ${easing}`;

    return {
      ...this.getExpandStyle(),
      transition,
    };
  }

  updateStatus = status => this.setState({ status });

  delay = (fn, time) => {
    this.timeout = setTimeout(fn, time);
  };

  clearDelay = () => {
    if (this.timeout) {
      clearTimeout(this.timeout);
    }
  };

  toggle = (open) => {
    // In firefox, setTimeout with duration 0 too short for browser notice the changes in dom
    const initialTransitDuration = 20;
    const { duration } = this.props;

    this.clearDelay();

    if (open) {
      this.updateStatus(STATUS.OPENING);

      this.delay(() => {
        this.updateStatus(STATUS.OPENED);

        this.delay(() => {
          this.updateStatus(STATUS.OPEN);
        }, duration);
      }, 0);
    } else {
      this.updateStatus(STATUS.CLOSING);

      requestAnimationFrame(() => {
        requestAnimationFrame(() => {
          this.updateStatus(STATUS.CLOSED);

          this.delay(() => {
            this.updateStatus(STATUS.CLOSE);
          }, duration);
        });
      });
    }
  };

  render() {
    const { className, children } = this.props;
    return (
      <div className={className} style={this.getStyle()} ref={(ref) => { this.refWrapper = ref; }}>
        {children}
      </div>
    );
  }
}

Expand.propTypes = {
  children: PropTypes.node.isRequired,
  open: PropTypes.bool,
  duration: PropTypes.number,
  easing: PropTypes.string,
  className: PropTypes.string,
};

Expand.defaultProps = {
  open: false,
  duration: 1000,
  easing: 'ease-in-out',
  className: '',
};

export default Expand;
