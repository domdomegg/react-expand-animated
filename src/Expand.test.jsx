import React from 'react';
import { render, mount } from 'enzyme';

import Expand from './Expand';

describe('Expand', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  it('when render expand with children --> then render default Expand with children', () => {
    const wrapper = render(<Expand>Hello</Expand>);

    expect(wrapper).toMatchSnapshot();
  });

  it('when render expand with open=true --> then render Expand with open styles', () => {
    const wrapper = render(<Expand open>Open true</Expand>);

    expect(wrapper).toMatchSnapshot();
  });

  it('when toggle open from false to true --> then change Expand styles from close to open', () => {
    const wrapper = mount(<Expand>Toggle</Expand>);

    expect(wrapper).toMatchSnapshot();

    wrapper.setProps({ open: true });

    jest.runOnlyPendingTimers();

    wrapper.update();

    expect(wrapper).toMatchSnapshot();

    jest.runOnlyPendingTimers();

    wrapper.update();

    expect(wrapper).toMatchSnapshot();

    expect(setTimeout).toHaveBeenCalledTimes(2);
  });

  it('when children is wrapped component --> then don\'t add new wrapper div', () => {
    const wrapper = mount(<Expand><p>Toggle</p></Expand>);

    expect(wrapper).toMatchSnapshot();

    wrapper.setProps({ open: true });

    jest.runOnlyPendingTimers();

    wrapper.update();

    expect(wrapper).toMatchSnapshot();

    jest.runOnlyPendingTimers();

    wrapper.update();

    expect(wrapper).toMatchSnapshot();

    expect(setTimeout).toHaveBeenCalledTimes(2);
  });

  it('when children is array --> then add a wrapper div', () => {
    const wrapper = mount(<Expand><p>Toggle 1</p><p>Toggle 2</p></Expand>);

    expect(wrapper).toMatchSnapshot();

    wrapper.setProps({ open: true });

    jest.runOnlyPendingTimers();

    wrapper.update();

    expect(wrapper).toMatchSnapshot();

    jest.runOnlyPendingTimers();

    wrapper.update();

    expect(wrapper).toMatchSnapshot();

    expect(setTimeout).toHaveBeenCalledTimes(2);
  });
});
