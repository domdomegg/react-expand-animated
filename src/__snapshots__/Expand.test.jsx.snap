// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expand when children is array --> then add a wrapper div 1`] = `
<Expand
  className=""
  duration={1000}
  easing="ease-in-out"
  open={false}
  tag="div"
>
  <div
    className=""
    style={
      Object {
        "height": 0,
        "opacity": 0,
        "overflow": "hidden",
        "transition": "height 1000ms ease-in-out, opacity 1000ms ease-in-out",
      }
    }
  >
    <p>
      Toggle 1
    </p>
    <p>
      Toggle 2
    </p>
  </div>
</Expand>
`;

exports[`Expand when children is array --> then add a wrapper div 2`] = `
<Expand
  className=""
  duration={1000}
  easing="ease-in-out"
  open={true}
  tag="div"
>
  <div
    className=""
    style={
      Object {
        "height": 0,
        "opacity": 1,
        "overflow": "hidden",
        "transition": "height 1000ms ease-in-out, opacity 1000ms ease-in-out",
      }
    }
  >
    <p>
      Toggle 1
    </p>
    <p>
      Toggle 2
    </p>
  </div>
</Expand>
`;

exports[`Expand when children is array --> then add a wrapper div 3`] = `
<Expand
  className=""
  duration={1000}
  easing="ease-in-out"
  open={true}
  tag="div"
>
  <div
    className=""
    style={
      Object {
        "height": "auto",
        "opacity": 1,
        "overflow": "unset",
        "transition": "height 1000ms ease-in-out, opacity 1000ms ease-in-out",
      }
    }
  >
    <p>
      Toggle 1
    </p>
    <p>
      Toggle 2
    </p>
  </div>
</Expand>
`;

exports[`Expand when children is wrapped component --> then don't add new wrapper div 1`] = `
<Expand
  className=""
  duration={1000}
  easing="ease-in-out"
  open={false}
  tag="div"
>
  <p
    style={
      Object {
        "height": 0,
        "opacity": 0,
        "overflow": "hidden",
        "transition": "height 1000ms ease-in-out, opacity 1000ms ease-in-out",
      }
    }
  >
    Toggle
  </p>
</Expand>
`;

exports[`Expand when children is wrapped component --> then don't add new wrapper div 2`] = `
<Expand
  className=""
  duration={1000}
  easing="ease-in-out"
  open={true}
  tag="div"
>
  <p
    style={
      Object {
        "height": 0,
        "opacity": 1,
        "overflow": "hidden",
        "transition": "height 1000ms ease-in-out, opacity 1000ms ease-in-out",
      }
    }
  >
    Toggle
  </p>
</Expand>
`;

exports[`Expand when children is wrapped component --> then don't add new wrapper div 3`] = `
<Expand
  className=""
  duration={1000}
  easing="ease-in-out"
  open={true}
  tag="div"
>
  <p
    style={
      Object {
        "height": "auto",
        "opacity": 1,
        "overflow": "unset",
        "transition": "height 1000ms ease-in-out, opacity 1000ms ease-in-out",
      }
    }
  >
    Toggle
  </p>
</Expand>
`;

exports[`Expand when render expand with children --> then render default Expand with children 1`] = `
<div
  class=""
  style="height:0;opacity:0;overflow:hidden;transition:height 1000ms ease-in-out, opacity 1000ms ease-in-out"
>
  Hello
</div>
`;

exports[`Expand when render expand with open=true --> then render Expand with open styles 1`] = `
<div
  class=""
  style="height:auto;opacity:1;overflow:unset;transition:height 1000ms ease-in-out, opacity 1000ms ease-in-out"
>
  Open true
</div>
`;

exports[`Expand when toggle open from false to true --> then change Expand styles from close to open 1`] = `
<Expand
  className=""
  duration={1000}
  easing="ease-in-out"
  open={false}
  tag="div"
>
  <div
    className=""
    style={
      Object {
        "height": 0,
        "opacity": 0,
        "overflow": "hidden",
        "transition": "height 1000ms ease-in-out, opacity 1000ms ease-in-out",
      }
    }
  >
    Toggle
  </div>
</Expand>
`;

exports[`Expand when toggle open from false to true --> then change Expand styles from close to open 2`] = `
<Expand
  className=""
  duration={1000}
  easing="ease-in-out"
  open={true}
  tag="div"
>
  <div
    className=""
    style={
      Object {
        "height": 0,
        "opacity": 1,
        "overflow": "hidden",
        "transition": "height 1000ms ease-in-out, opacity 1000ms ease-in-out",
      }
    }
  >
    Toggle
  </div>
</Expand>
`;

exports[`Expand when toggle open from false to true --> then change Expand styles from close to open 3`] = `
<Expand
  className=""
  duration={1000}
  easing="ease-in-out"
  open={true}
  tag="div"
>
  <div
    className=""
    style={
      Object {
        "height": "auto",
        "opacity": 1,
        "overflow": "unset",
        "transition": "height 1000ms ease-in-out, opacity 1000ms ease-in-out",
      }
    }
  >
    Toggle
  </div>
</Expand>
`;
